File: synara/src/app/features/matrix-dto/notification.ts
/**
 * OS notification candidate DTO — privacy-filtered title/body only.
 */

import type { EventId, NotificationCandidateId, RoomId } from './ids';
import {
  hasForbiddenWireFields,
  isObject,
  optString,
  reqBoolean,
  reqString,
} from './parseUtil';

export const NOTIFICATION_KINDS = [
  'message',
  'invite',
  'agent_approval',
  'later_reminder',
] as const;
export type NotificationKind = (typeof NOTIFICATION_KINDS)[number];
const KIND_SET = new Set<string>(NOTIFICATION_KINDS);

export function isNotificationKind(value: unknown): value is NotificationKind {
  return typeof value === 'string' && KIND_SET.has(value);
}

export type NotificationCandidate = {
  candidateId: NotificationCandidateId;
  roomId: RoomId;
  eventId?: EventId;
  kind: NotificationKind;
  title: string;
  body: string;
  route?: string;
  suppressIfFocusedRoom: boolean;
  isEncrypted: boolean;
};

export function parseNotificationCandidate(
  value: unknown
): NotificationCandidate | null {
  if (!isObject(value) || hasForbiddenWireFields(value)) return null;
  const candidateId = reqString(value, 'candidateId');
  const roomId = reqString(value, 'roomId');
  const eventId = optString(value, 'eventId');
  const title = reqString(value, 'title');
  const body = reqString(value, 'body');
  const route = optString(value, 'route');
  const suppressIfFocusedRoom = reqBoolean(value, 'suppressIfFocusedRoom');
  const isEncrypted = reqBoolean(value, 'isEncrypted');
  if (
    candidateId === null ||
    roomId === null ||
    eventId === null ||
    title === null ||
    body === null ||
    route === null ||
    suppressIfFocusedRoom === null ||
    isEncrypted === null ||
    !isNotificationKind(value.kind)
  ) {
    return null;
  }
  return {
    candidateId,
    roomId,
    eventId,
    kind: value.kind,
    title,
    body,
    route,
    suppressIfFocusedRoom,
    isEncrypted,
  };
}

File: src-tauri/src/matrix/dto/notification.rs
//! OS notification candidate DTO — privacy-filtered title/body only.

use serde::{Deserialize, Serialize};

use super::ids::{EventId, NotificationCandidateId, RoomId};

/// Product notification kind (maps to desktop notification routing).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum NotificationKind {
    Message,
    Invite,
    AgentApproval,
    LaterReminder,
}

impl NotificationKind {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Message => "message",
            Self::Invite => "invite",
            Self::AgentApproval => "agent_approval",
            Self::LaterReminder => "later_reminder",
        }
    }
}

/// Candidate for an OS-level desktop notification.
///
/// `title` / `body` are already privacy-filtered product strings — never raw
/// ciphertext or full event dumps.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NotificationCandidate {
    pub candidate_id: NotificationCandidateId,
    pub room_id: RoomId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub event_id: Option<EventId>,
    pub kind: NotificationKind,
    pub title: String,
    pub body: String,
    /// Product deep-link route (e.g. `/home/room/!id:server`).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub route: Option<String>,
    pub suppress_if_focused_room: bool,
    pub is_encrypted: bool,
}
